scripts/hashauthservice

```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+
+import csv
+from functools import wraps
+import os
+from urllib.parse import urlparse
+
+import requests
+from traitlets.config import PyFileConfigLoader
+
+from tornado.ioloop import IOLoop
+from tornado.httpserver import HTTPServer
+from tornado.web import RequestHandler, Application, authenticated, HTTPError
+
+from jupyterhub.services.auth import HubOAuthenticated, HubOAuthCallbackHandler
+from jupyterhub.utils import url_path_join
+
+from hashauthenticator import generate_password_digest
+
+service_url = urlparse(os.environ['JUPYTERHUB_SERVICE_URL'])
+prefix = os.environ.get('JUPYTERHUB_SERVICE_PREFIX', '/')
+api_token = os.environ['JUPYTERHUB_API_TOKEN']
+config_file = os.environ.get('CONFIG_FILE', 'jupyterhub_config.py')
+
+
+def generate_password_function():
+  config = PyFileConfigLoader(config_file).load_config()
+  hash_config = config.get('HashAuthenticator', dict())
+  secret_key = hash_config.get('secret_key', '')
+  password_length = hash_config.get('password_length', 6)
+  return lambda user: generate_password_digest(user, secret_key, password_length)
+
+get_password = generate_password_function()
+
+
+class HashAuthHandler(HubOAuthenticated, RequestHandler):
```

Comment:
Is it worth using flask instead?  I'm thinking about keeping our development compatible with our main website.  Don't know if it's worth rewriting this though ...

Reply:
If you look back in the git history, you'll see I started with a Flask version.  I think tornado is right for this particular project, though.  Since we've released this as an open-source project, we should make it as easy to install as possible.  If they have JupyterHub, they have Tornado, so we might as well just use that.  JupyterHub provides some authentication helpers for tornado, so this code is actually more compact than the Flask version.

Also, I don't see this expanding past a single endpoint.  The grader, on the other hand, I've done in Flask and I see no reason to move it to tornado.